# Solving three problems from the Euler Project (with increasing difficulty)

### Problem 5: Smallest Multiple  
  
Description: 2520 is the smallest number that can be divided by each of the numbers from 1 to 10 without any remainder.  
What is the smallest positive number that is evenly divisible by all of the numbers from 1 to 20?   
  

After reading this question, I immediately thought about prime factorization and factor trees. The goal is to program a factor tree and work backwards from the prime factors to calculate the smallest number that is evenly divisible by all numbers from 1 to 20.    
Note that 2520 has to be divisible by the prime numbers from 1 to 10 in order to be divisible by all numbers 1 to 10 without any remainder. Knowing this, we don't have to check whether 2520 is divisible by every number from 1 to 10 and this will save us time.  
So to begin, we will define a function to return a list of all the prime numbers from 1 to a given number. This will help us code our factor tree.


In [60]:
def prime_num(n):
    """
    return list of prime numbers given a number
    """
    primes = []
    for number in range(n + 1):
        if number > 1:
            # prime numbers have to be greater than 1
            for i in range(2, number):
                if (number % i) == 0:
                    # number is not prime (number is divisible by a number other than one and itself)
                    # so, return to outer for loop and increment by one
                    break
            else:
                # number is prime so add to list
                primes.append(number)
    return primes

# print(prime_num(10))

[2, 3, 5, 7]


Next, the goal is to code a factor tree where the "end" or what is returned are all the prime factors of the given number.   
We will find the prime factors of this given number (n) using knowledge about whether a number is prime in range 1 to n from the above function that returns the prime numbers in range 1 to n.

In [61]:
def find_prime_factors(n):
    """
    recursively find prime factors of number given prime numbers in range 1 to that number
    """
    if n == 1:
        # base case: if n = 1, there are no prime factors
        return []
    for num in prime_numbers:
        # loop through each number in the list of prime numbers in range 1 to n
        if n % num == 0:
            # return the prime factor and use recursion to find next prime factor
            return [num] + find_prime_factors(n / num)

[2, 2, 5]


We will write a final function that uses the two functions we have written. We know the prime numbers from 1 to a given number from the first function. We can also find the prime factors of a number from the second function.  

Now, we will find the prime factors of every number below 20 (using the second function). We will keep track of the maximum prime factors. For example, the prime factors of 2 and 3 respectively are 2 and 3 and the prime factors of 4 are 2 and 2, so the maximum count of the prime factor 2 is 2 and of 3 is 1 and the prime factors of the smallest number evenly divisible by 2, 3, and 4, are 2, 2, and 3 and NOT 2, 3, 2, and 2. This means the smallest number equals 2*2*3 = 12.    

Using this same logic, we will use these prime factors to calculate the smallest number, where the smallest number has the same prime factors.  
 
Since the problem asks us to find the smallest positive number that is evenly divisible by all of the numbers from 1 to 20, we will assume that n is 20 and that the range is from 1 to 20.  

We recently reviewed the collections module, so we will use that to count the maximum number of each prime factor.

In [81]:
from collections import Counter

def main():
    """
    print smallest number evenly divisible by all numbers from 1 to 20
    """
    n = 20
    prime_numbers = prime_num(n)
    max_primes = Counter()
    
    for i in range(2, 21):
        # creating a factor tree for each number from 2 to 20
        prime_factors = Counter(find_prime_factors(i))
        # keeps maximum of counts of each prime factor
        max_primes = prime_factors | max_primes
        
    smallest_num = 1
    for prime, count in max_primes.items():
        # based on prime factors, calculate smallest number by multiplying
        smallest_num = smallest_num * (prime ** amount)
    
    print(smallest_num)

main()

232792560


The smallest positive number that is evenly divisible by all of the numbers from 1 to 20 is 232792560.

### Problem 39: Integer Right Triangles  
  
Description: If p is the perimeter of a right angle triangle with integral length sides, {a,b,c}, there are exactly three solutions for p = 120.  
{20,48,52}, {24,45,51}, {30,40,50}  
For which value of p ≤ 1000, is the number of solutions maximised?

This question involves pythagorean theorem ($a^2 + b^2 = c^2$). The first step is to define a function that returns the number of solutions that meet the right angle triangle side length definition given a perimeter.  

We know that a and b have to be at least 1 and cannot be more than half the perimeter. So, we will use a to loop through all integers from 1 to $p/2$.

Given these two equations:  
$a^2 + b^2 = c^2$  (pythagorean theorem)   
$a + b + c = p$  (definition of perimeter)

We can solve for b and c.  
1) $a + b + c = p$ --> $b+c = p-a$  
2) $a^2 + b^2 = c^2$ --> $c-b = \frac{a^2}{p-a}$  
3) $(c-b) + (c+b) = \frac{a^2}{p-a} + (p-a)$ --> $c = \frac{a^2 + (p-a)^2}{2 *(p-a)}$

Then, using pythagorean theorem, $b = \sqrt{c^2 - a^2}$  

The final step is to check whether the definition of perimeter is met and make sure we have the correct value for a. 

In [91]:
def find_side_lengths(p):
    """
    return number of solutions to right angle triangles given a perimeter
    """
    num_solutions = 0
    
    # an integer side length of a right triangle has to be at least 1 and not more than half the perimeter
    for a in range(1, p//2):
        # solve pythagorean theorem for c and b
        c = (a**2 + (p-a)**2) // (2 * (p-a))
        b = (c**2 - a**2) ** 0.5
        
        # definition of perimeter
        if a + b + c == p:
            num_solutions += 1
    
    # a and b can be switched in a solution, but we only want one solution of possible integral side lengths
    return num_solutions//2

Next, we want to find the perimeter where the number of solutions, given that we know the number of solutions for a perimeter from the previous exquation, is maximized for perimeters ≤ 1000. We will use a dictionary to store the perimeter and the number of solutions for that perimeter. 

In [93]:
def maximized(perimeter):
    """
    print perimeter where the number of solutions is maximized for perimeter <= 1000 
    """
    data = {"solutions": 0, "perimeter": 0}
    for p in range(perimeter + 1):
        if find_side_lengths(p) > data['solutions']:
            data['solutions'] = find_side_lengths(p)
            data['perimeter'] = p
    return data['perimeter']

print(maximized(1000))

840


The number of solutions is maximised at a perimeter of 840. 

### Problem 112: Bouncy Numbers  
  
Decription: Working from left-to-right if no digit is exceeded by the digit to its left it is called an increasing number; for example, 134468.  
  
Similarly if no digit is exceeded by the digit to its right it is called a decreasing number; for example, 66420.  
  
We shall call a positive integer that is neither increasing nor decreasing a "bouncy" number; for example, 155349.  
  
Clearly there cannot be any bouncy numbers below one-hundred, but just over half of the numbers below one-thousand (525) are bouncy. In fact, the least number for which the proportion of bouncy numbers first reaches 50% is 538.  
  
Surprisingly, bouncy numbers become more and more common and by the time we reach 21780 the proportion of bouncy numbers is equal to 90%.  
  
Find the least number for which the proportion of bouncy numbers is exactly 99%.  



In [ ]:
def bouncy(n):
    """
    return True if the number is a bouncy number and false if the number is not bouncy
    """
    number = list(map(int, str(n))) # create list of digits
    increasing = sorted(number)
    decreasing = increasing[::-1]
    return increasing != number and decreasing != number